### ISSUE TITLE
Title input overflow

### ISSUE THREAD

**Issue by JohnONolan** [TryGhost/Ghost-Editor#9]
Currently the post title is a text input, which means it overflows horizontally. Ideally this should just behave like normal text and break to a new line.

**Comment by disordinary** [comment]
I've got a working and reliable solution, the caveat is that the title is now part of the post content and not a separate field. 

I've added helper functions to extract and replace the content text of the title so that Ghost can use it for structured data, however right now the title will appear within the content block in templates. 

We do have a few options:
- We can modify the title helper, excerpt helper, and content helper so that they pull the right content out of the block of rendered HTML or AMP.
- We can fork the mobiledoc HTML renderer so that it doesn't draw the first H1.
- We can leave the title in the content block and get rid of the title helper within the theme API.

**Comment by ErisDS** [comment]
Is there somewhere an explanation of what the desired UX for the title is and why this is so hard to achieve?

Both this issue and the convo in Slack issue jumps from needing a line wrap in the title field to suddenly having to implement really awkward workarounds all over the place, and I'm not able to follow the reasoning.

**Comment by kevinansfield** [comment]
Agreed, the normal solution for the UI that @JohnONolan has described is to change the text input to a textarea and have it auto-expand to fit the content - is there some other reason we need to move the title into the mobiledoc editor?

**Comment by disordinary** [comment]
Visually it's part of the editor, because of that a user expects:
- To be able to press the down arrow and move from the title into the main body.
- To be able to press the up arrow from the body and to go back to the title.
- To be able to press enter within the title and have it split into the body.
- To be able to press backspace within the body and have that block join with the title.
- To be able to style the title.
- To be able to undo and redo the title as if it's the body.

**Comment by disordinary** [comment]
If the title and body were visually distinct and it was clear that they weren't the same interface then these wouldn't be issues. But because they're combined and people are used to the Medium editor I think they need to be integrated. Because they're combined visually users will expect a integrated experience.

**Comment by disordinary** [comment]
I also have a reliable and intuitive solution for mobiledoc but it will require a bit of a rework of the admin UI and template helpers as discussed above.

**Comment by disordinary** [comment]
Cory and Mathew from mobiledoc suggested another solution to this, and that is embedding different mobiledoc containers within one mobiledoc document, the first for the title, the second for the content.

This solves the issue with having seamless cursor movement between the title and body and has them as independent entities but has a few trade-offs.
- You cannot select between the title and the body.
- There is not one undo queue.
- Pressing enter in the title doesn't split it and generate a new paragraph in the body.

IMHO having the title inside the body is still the best solution from an editing experience.

**Comment by ErisDS** [comment]
Ok so, I understand the reason why it's fiddly to merge 2 UI components seamlessly, however, I really feel like I'm missing something when it comes to this part:

> I've added helper functions to extract and replace the content text of the title so that Ghost can use it for structured data, however right now the title will appear within the content block in templates.

I don't understand why, if we have to process the data on save to pull the title out into the `post.title` field, why we can't at the same point also not add the title to the `post.html` field?

**Comment by disordinary** [comment]
It should be trivial to remove the title from the body, but it has to happen in one of three ways IMHO.
1. We have a post render step after the mobiledoc is generated to remove the title from the HTML/AMP/plaintext (or whatever), or
2. We fork each of the necessary mobiledoc renderers so that we can remove the title as it's rendered, or
3. We modify the mobiledoc on save to pull out the title and load to replace it.

None of these feel like great solutions, although 3 is probably best. But. instead of working with the editor though (which relies on the cursor range object in the DOM), we'd be better placed if we put together a library for manipulating mobiledoc documents outside of a DOM situation. This could be very useful for other things so it's worth investigating regardless. 

To be clearer I wasn't talking about pulling the content out of the mobiledoc on save but rather just getting the plain text version of the first block element to use for our metadata rather than anything more invasive to the document.

**Comment by ErisDS** [comment]
Can you please point me at the piece of code which takes the existing mobiledoc document and converts it to HTML to store in the database?

**Comment by disordinary** [comment]
https://github.com/TryGhost/Ghost/blob/2d19ae2c6c403438c714f880ff56bd7555cd8583/core/server/models/post.js#L207

`converter` is an instance of the mobiledoc HTML renderer which has been setup with our cards, atoms, and various other options. 

As you can see it's reasonably straight forwards ATM and works almost identically to how showdown works.

**Comment by ErisDS** [comment]
So... as I see, the existing solution goes from JSON -> HTML string. We could regex parse this before storing it, given that we know the structure, but regex parsing of HTML is always GROSS.

However, it seems to me this method of getting from Mobiledoc -> HTML is deprecated.

The new method [documented here](https://github.com/bustlelabs/mobiledoc-dom-renderer) involves a middle step where we have access to a SimpleDOM object. If we update to use this new method, we have a point at which we can manipulate the DOM.

This would allow us to get rid of the H1 element easily?

Also, we could get rid of that awkward wrapping `<div></div>`, if there isn't another way to configure mobiledoc to not do that?!

**Comment by disordinary** [comment]
Oh nice, they've obviously deprecated it since we wrote the front-end code. using simpledom should make life easier. 

The question then is do we want to have the title be formatted?

**Comment by ErisDS** [comment]
It isn't right now. I don't know if there's any plan or desire to add that @JohnONolan ?

**Comment by disordinary** [comment]
If it's any consolation it's going to be easier to keep the formatting in mobiledoc than not (we'd have to disable the toolbar and formatting hotkeys if the title is selected, and it's going to be even more of a pain if the user selects the title and content and wants to format it).

**Comment by JohnONolan** [comment]
sorry about being late to this! I have some definite simplifications in requirements here which might help?

- ~~To be able to press enter within the title and have it split into the body.~~
- ~~To be able to press backspace within the body and have that block join with the title.~~
- ~~To be able to style the title.~~
- ~~To be able to undo and redo the title as if it's the body.~~

Definitely not needed. These are edge cases rather than use-cases!

**Comment by disordinary** [comment]
Thanks John,

How about selecting between the title and body?

Anyway, with the above requirements it would appear that a nested mobiledoc solution might be the best option - I have no idea how to implement that though so will need to do a bit of research.

If we go with a textarea for the title then navigating between elements with the up and down keys is going to be challenging. 

I'm not convinced that these are edge cases though but rather expected behavior.

**Comment by JohnONolan** [comment]
Selecting between the title + body would be nice yeah, to copy/paste a whole post out.

> I'm not convinced that these are edge cases though but rather expected behavior.

But when you apply the logic of "hey that doesn't work" to each of those points above, the obvious answer is "well don't do that then" -- it's really not a thing that people NEED to be able to do. It's not a valid feature or thing that is required. 

Like "ah fuck howcome I can't press enter in the middle of my title and have it automatically move to the body" -- why are you even trying to do that? And why should that be a case which is planned for? If anything I would expect pressing enter in the middle of a title to create a line-break within the title itself - which is what happens in most editors

**Comment by disordinary** [comment]
The only way to be able to select the text in both the title and body is to have everything in one mobiledoc document. 😞

**Comment by JohnONolan** [comment]
I don't think that one is a dealbreaker though. It would be nice - yes - but if that's the ONLY reason to combine the fields, then it's not a good enough reason?

**Comment by disordinary** [comment]
So the only requirement is to be able to navigate with the up and down keys between the body and the title or is that also not a requirement?

**Comment by JohnONolan** [comment]
The measurement of what's a requirement, what's not, and what's nice to have, is based pretty much entirely on the frequency of use that a feature will get:

Keying up/down from title to body: Very common. Probably done on every *single* post written in the editor either via arrow keys or by enter key. == requirement

Copying and pasting all text+title in/out of the editor: Common-ish? Probably done by some people regularly and some people never at all. Having to copy title/body separately is a mild inconvenience for a non-primary usecase. == nice to have, not essential

Splitting title into body / joining body onto title: Rare. You really only do this if you've already made a mistake, or you end up in some weird situation where you somehow need to do this. This isn't a planned state which you will definitely end up in each time you use the editor. == edge case, not important

With everything like this, my question is: How often are people going to run into it? Is this something they need to inherently do with every post? or is it an edge case which is needed sometimes?

**Comment by disordinary** [comment]
Cool, so the solution is the "nested mobiledocs" solution in the editor and then seperate the fields when they're saved in ghost-admin.

The problem is I have no idea how to do this as it was only ever mentioned casually in a conversation. I'll follow up with Cory and Mathew (who have both indicated that they're absolutely flat out for the next two weeks) to get pointed in the right direction.
